### PR TITLE
Feature/quit household

### PR DESCRIPTION
**Pull request by Oxydeas**
**Features**

- Possibility to quit a household as aresident
- Possibility of in app scanning of QR code to join a household

This code is tested as much as I could, since we cannot mimic the camera scanning something

### CODE REVIEW

File: app/src/androidTest/java/com/github/houseorganizer/houseorganizer/HouseSelectionActivityTestLeave.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        Long num_residents_after = (Long) houseData_after.get("num_members");
+
+        assertTrue(resident_before.contains(TEST_USERS_EMAILS[1]));
+        assertFalse(resident_after.contains(TEST_USERS_EMAILS[1]));
+        Long expected = num_residents_before - 1;
+        assertEquals(expected, num_residents_after);
+
+        FirebaseTestsHelper.createHouseholds();
+    }
+}
```
Comment (Arkadyus) [inline app/src/androidTest/java/com/github/houseorganizer/houseorganizer/HouseSelectionActivityTestLeave.java:88]: can this not be moved in HouseSelectionActivityTest ?

File: app/src/main/java/com/github/houseorganizer/houseorganizer/shop/FirestoreShopList.java [hunk 2]
```diff
@@ -112,12 +101,26 @@ private static List<Map<String, Object>> convertItemsListToFirebase(List<ShopIte
...
+            return Tasks.forCanceled();
+
+        return onlineReference.get().continueWith( r -> {
+            if(!r.isSuccessful())
+                return null;
+            DocumentSnapshot snap = r.getResult();
+            setItems(convertFirebaseListToItems((List<Map<String, Object>>) snap.get("items")));
+            return snap;
+        });
+    }
```
Comment (Arkadyus) [inline app/src/main/java/com/github/houseorganizer/houseorganizer/shop/FirestoreShopList.java:119]: from what i found this is only used in testing, does it make sense to have it ?

**Review by Arkadyus — CHANGES_REQUESTED** [review]
Everything looks fine, just a few questions i commented about

File: app/src/androidTest/java/com/github/houseorganizer/houseorganizer/HouseSelectionActivityTestLeave.java (hunk 1 above)
Comment (Oxydeas) [reply app/src/androidTest/java/com/github/houseorganizer/houseorganizer/HouseSelectionActivityTestLeave.java:86]: Sadly not as the device remembers user 1 as owner, it is a strange behaviour but I couldn't make it work

File: app/src/main/java/com/github/houseorganizer/houseorganizer/shop/FirestoreShopList.java (hunk 2 above)
Comment (Oxydeas) [reply app/src/main/java/com/github/houseorganizer/houseorganizer/shop/FirestoreShopList.java:119]: I don't know. I copy pasted the fix from Sandra in order to finally get my tests passing

File: app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.google.zxing.BinaryBitmap;
+import com.google.zxing.PlanarYUVLuminanceSource;
+import com.google.zxing.Result;
+import com.google.zxing.common.HybridBinarizer;
+import com.google.zxing.multi.qrcode.QRCodeMultiReader;
+
+import java.nio.ByteBuffer;
+
+public class QRAnalyzer implements ImageAnalysis.Analyzer {
```
Comment (aindreias) [inline app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java:18]: Could you make this class final & the same for its listener attribute?

File: app/src/main/java/com/github/houseorganizer/houseorganizer/house/QRCodeScanActivity.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        db = FirebaseFirestore.getInstance();
+        auth = FirebaseAuth.getInstance();
+
+        cameraPreview = findViewById(R.id.cameraPreview);
+        cameraProvider = ProcessCameraProvider.getInstance(this);
+
+        askForCameraPermissionsOrContinue();
+    }
+
+    private void askForCameraPermissionsOrContinue() {
```
Comment (aindreias) [inline app/src/main/java/com/github/houseorganizer/houseorganizer/house/QRCodeScanActivity.java:62]: Nice method - will probably be using something similar for the offline mode

**Review by aindreias — CHANGES_REQUESTED** [review]
Great work, I only have a few comments / questions. Approving once they're answered 👍

**Review by Arkadyus — APPROVED** [review]
Thanks for the answers everything looks clear now

File: app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java (hunk 3 above)
Comment (Oxydeas) [reply app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java:18]: done

File: app/src/main/java/com/github/houseorganizer/houseorganizer/house/QRCodeScanActivity.java (hunk 4 above)
Comment (Oxydeas) [reply app/src/main/java/com/github/houseorganizer/houseorganizer/house/QRCodeScanActivity.java:62]: Yup, seems to work

File: app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java (hunk 3 above)
Comment (aindreias) [reply app/src/main/java/com/github/houseorganizer/houseorganizer/image/QRAnalyzer.java]: Nice, thanks!

**Review by aindreias — APPROVED** [review]
Thank you for the changes! Seems all good to me!

**Pull request merged**